src/middlewared/middlewared/plugins/filesystem.py

```diff
@@ -476,21 +484,17 @@ def file_receive(self, path, content, options):
 
         `content` must be a base 64 encoded file content.
         """
-        dirname = os.path.dirname(path)
-        if not os.path.exists(dirname):
-            os.makedirs(dirname)
-        if options.get('append'):
-            openmode = 'ab'
-        else:
-            openmode = 'wb+'
-        with open(path, openmode) as f:
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path, 'ab' if options.get('append') else 'wb+') as f:
             f.write(binascii.a2b_base64(content))
-        if mode := options.get('mode'):
-            os.chmod(path, mode)
-        if (uid := options.get('uid')) is not None and (gid := options.get('gid')) is not None:
-            os.chown(path, uid, gid)
+            if mode := options.get('mode'):
+                os.fchmod(f.fileno(), mode)
+
+            os.fchown(f.fileno(), options.get('uid', -1), options.get('gid', -1))
+
         if path == PWENC_FILE_SECRET:
```

Comment:
I wonder if in case of PWENC_FILE_SECRET we should always force mode to be 0o600 (avoid potential for someone writing bug where they make it world-readable).

Reply:
Agree, I actually kind of assumed that the api call below this check did so. I'll make the change.